### PR TITLE
feat(parquet): Support LIST merge and sorting

### PR DESCRIPTION
**Pull request by linuxpi**
### Description
Promote compatible scalar and LIST schemas during Parquet merge and wrap scalar values as singleton lists. Derive the output Parquet schema from the canonical Arrow schema.

Sort LIST fields by the minimum non-null element in both the chunk writer and k-way merge while preserving original list values and null versus empty distinctions.

Add mixed-generation, null-placement, multi-column, deferred-read, and chunked-writer coverage.

NOTE: Currently the PR doesn't integrate `index.sort.mode` and assumed default `MIN`. A followup PR will address this

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: sandbox/plugins/parquet-data-format/src/main/rust/src/merge/heap.rs [hunk 1]
```diff
@@ -201,6 +268,106 @@ pub fn get_sort_value(
...
+                )
+            }
+        })
+        .collect::<MergeResult<Vec<_>>>()?;
+    Ok(take(values.as_ref(), &UInt64Array::from(indices), None)?)
+}
+
+/// Returns the scalar type used as the physical sort key. LIST columns always
+/// reduce to the natural minimum non-null element; no customer-selectable mode
+/// is exposed by the Parquet writer.
```
Comment (Shailesh-Kumar-Singh) [inline sandbox/plugins/parquet-data-format/src/main/rust/src/merge/heap.rs]: We have `index.sort.mode` [can be min / max ]  index setting also which controls what value is picked to sort the document for multi values fields.
why are we hardcoding the mode to `min` here?
Even if we do not specify the sort mode, if sort order is descending, then the mode defaults to max, So will this not be a deviation from the current Opensearch behaviour?

Comment (linuxpi) [reply sandbox/plugins/parquet-data-format/src/main/rust/src/merge/heap.rs]: you are right @Shailesh-Kumar-Singh . We will integrate those modes in followup PRs. I should have mentioned this in the description. will update.

File: sandbox/plugins/parquet-data-format/src/main/rust/src/merge/heap.rs [hunk 2]
```diff
@@ -125,27 +133,69 @@ impl PartialOrd for HeapItem {
...
+    end: usize,
+    dtype: &ArrowDataType,
+    null_first: bool,
+) -> MergeResult<SortKey> {
+    let mut minimum = None;
+    for index in start..end {
+        if values.is_null(index) {
+            continue;
+        }
+        let candidate = get_array_sort_value(values, index, dtype, false)?;
```
Comment (Shailesh-Kumar-Singh) [inline sandbox/plugins/parquet-data-format/src/main/rust/src/merge/heap.rs]: This builds a full SortKey per candidate before comparing, so Utf8/LargeUtf8/Binary/LargeBinary allocate a `Vec<u8> (.to_vec())` for every element and discard all but the min. 
Can we just instead scan for the winning index, comparing in place and build one SortKey at the end (primitives already don't allocate so that can stay as it is)

File: sandbox/plugins/parquet-data-format/src/main/rust/src/merge/heap.rs [hunk 3]
```diff
@@ -201,6 +268,106 @@ pub fn get_sort_value(
...
+    start: usize,
+    end: usize,
+    dtype: &ArrowDataType,
+) -> MergeResult<Option<u64>> {
+    let mut minimum = None;
+    for index in start..end {
+        if values.is_null(index) {
+            continue;
+        }
+        let candidate = get_array_sort_value(values, index, dtype, false)?;
```
Comment (Shailesh-Kumar-Singh) [inline sandbox/plugins/parquet-data-format/src/main/rust/src/merge/heap.rs]: Same allocation issue as `get_min_value`. Once the byte types compute a winning index anyway (based on the comment above), can we collapse them into a single index-finding helper?
Here since the function only needs the index, comparing as_string()/as_binary().value(i) in place would find the winner with zero allocations.

File: sandbox/plugins/parquet-data-format/src/main/rust/src/merge/context.rs [hunk 4]
```diff
@@ -52,12 +53,59 @@ pub struct MergeContext {
...
+    })
+}
+
 impl MergeContext {
     /// Creates a new merge context: builds union schemas, opens the output
     /// writer, and spawns the background IO task.
     pub fn new(
         arrow_schemas: Vec<ArrowSchema>,
-        parquet_descriptors: &[SchemaDescriptor],
+        _parquet_descriptors: &[SchemaDescriptor],
```
Comment (Shailesh-Kumar-Singh) [inline sandbox/plugins/parquet-data-format/src/main/rust/src/merge/context.rs]: This parameter is unused now, can we remove this? 
Both call sites still build and clone a `Vec<SchemaDescriptor>` to pass it

File: sandbox/plugins/parquet-data-format/src/main/rust/src/merge/schema.rs [hunk 5]
```diff
@@ -130,7 +110,37 @@ impl ColumnMapping {
...
+                            )));
+                        }
+                        let offsets =
+                            OffsetBuffer::new((0..=num_rows as i32).collect::<Vec<_>>().into());
+                        columns.push(Arc::new(ListArray::new(
+                            Arc::clone(child),
+                            offsets,
+                            source.clone(),
+                            source.nulls().cloned(),
+                        )));
```
Comment (LantaoJin) [inline sandbox/plugins/parquet-data-format/src/main/rust/src/merge/schema.rs:134]: Can you add a test with a null scalar and a non-nullable LIST child? `ListArray::new` panics when the LIST child is non-nullable but the scalar source contains nulls.

Comment (linuxpi) [reply sandbox/plugins/parquet-data-format/src/main/rust/src/merge/schema.rs]: @LantaoJin we went back on the auto promotion thing for now. we would be initially going with strict multi-value true or false from the very beginning of index creation.

File: sandbox/plugins/parquet-data-format/src/main/rust/src/merge/heap.rs [hunk 6]
```diff
@@ -125,27 +133,153 @@ impl PartialOrd for HeapItem {
...
     }
+    // `max` only affects LIST/FixedSizeList reduction (MIN vs MAX element). For
+    // scalar columns the value at `row` is the sort key regardless of mode.
     let key = match dtype {
-        // Integer types → SortKey::Int
         ArrowDataType::Int64 => SortKey::Int(col.as_primitive::<Int64Type>().value(row)),
         ArrowDataType::Int32 => SortKey::Int(col.as_primitive::<Int32Type>().value(row) as i64),
         ArrowDataType::Int16 => SortKey::Int(col.as_primitive::<Int16Type>().value(row) as i64),
         ArrowDataType::Int8 => SortKey::Int(col.as_primitive::<Int8Type>().value(row) as i64),
+        ArrowDataType::UInt64 => SortKey::UInt(col.as_primitive::<UInt64Type>().value(row)),
```
Comment (linuxpi) [inline sandbox/plugins/parquet-data-format/src/main/rust/src/merge/heap.rs:279]: some of these newly added DataType are not supported by Opensearch, will remove those

File: sandbox/plugins/parquet-data-format/src/main/rust/src/merge/heap.rs (hunk 3 above)
Comment (linuxpi) [reply sandbox/plugins/parquet-data-format/src/main/rust/src/merge/heap.rs]: done

File: sandbox/plugins/parquet-data-format/src/main/rust/src/merge/heap.rs (hunk 2 above)
Comment (linuxpi) [reply sandbox/plugins/parquet-data-format/src/main/rust/src/merge/heap.rs]: done

File: sandbox/plugins/parquet-data-format/src/main/rust/src/merge/context.rs (hunk 4 above)
Comment (linuxpi) [reply sandbox/plugins/parquet-data-format/src/main/rust/src/merge/context.rs]: done

File: sandbox/plugins/parquet-data-format/src/main/rust/tests/merge_list_column_tests.rs [hunk 7]
```diff
@@ -0,0 +1,898 @@
...
+    assert_eq!(pairs[0], (1, v(&["beta", "alpha", "beta"])));
+    assert_eq!(pairs[1], (2, None), "null list must stay null");
+    assert_eq!(pairs[2], (3, v(&["solo"])));
+    assert_eq!(pairs[3].0, 4);
+    assert_eq!(pairs[3].1, Some(vec![]), "empty list must stay empty");
+    assert_eq!(pairs[4], (5, v(&["x", "y"])));
+}
+
+#[test]
+fn unsorted_merge_promotes_scalar_values_to_singleton_lists() {
```
Comment (rayshrey) [inline sandbox/plugins/parquet-data-format/src/main/rust/tests/merge_list_column_tests.rs]: Both `unsorted_merge_promotes_scalar_values_to_singleton_lists` and `sorted_merge_orders_scalar_and_list_generations_by_minimum_value` tests will fail as merges of scalar and list values are not yet supported. We might need to remove these tests.

On investigation looks like the Rust side unit tests are not a part of the sandbox check and hence it wasn't caught, we might need to fix this as well.

**Review by rayshrey — COMMENTED** [review]
Can we have merge specific ITs for List related fields as part of the Composite ITs that we have. Good to cover cases like merges are successful with list fields, and fail with auto promotions(can revisit the auto promotion ITs in the feature flag gating PR)

File: sandbox/plugins/parquet-data-format/src/main/rust/tests/merge_list_column_tests.rs (hunk 7 above)
Comment (linuxpi) [reply sandbox/plugins/parquet-data-format/src/main/rust/tests/merge_list_column_tests.rs]: thanks @rayshrey , i wasn't aware rust tests are not run. will run them locally for now to validate

Comment (linuxpi) [reply sandbox/plugins/parquet-data-format/src/main/rust/tests/merge_list_column_tests.rs]: Removed the irrelevant tests

Comment (linuxpi) [reply sandbox/plugins/parquet-data-format/src/main/rust/tests/merge_list_column_tests.rs]: Added ParquetListSortModeIntegrationTests for merge coverage

**Review by rayshrey — APPROVED** [review]
